### ISSUE TITLE
Explanation of how this template works

### ISSUE THREAD

**Issue by bigdav1178** [fHDHR/fHDHR_Template#38]
Can you explain how this template works: what info it is looking for, and how should that info should be passed to the script? How is the end user selecting the desired stream, and what gets passed back to the user (the originating stream URL?)?

I'm trying to come up with a variant that connects to YouTube TV. So far I've successfully navigated authentication, scraped channel info and generated streams (m3u8's). I figure I can put all this together and pass it into your template, I'm just trying to figure out how.

At this time, I'm not concerned with EPG; that work is secondary to making the streams work. If you can pass along how to disable the EPG portion of the template, that would also help. Thanks!

**Comment by deathbybandaid** [comment]
I just got back home from a vacation, and I'll give you a deep dive reply when I get settled.

I'd suggest the youtube varient to look at what it does since it makes good use of youtube api.

If you aren't concerned with epg, return an empty dictionary. (Returning no program data in the dict will generate 7 days of half hour blocks)

**Comment by bigdav1178** [comment]
I looked into the youtube variant (specifically, the prayer house one as an example), and I may be able to make it work. YouTube TV uses a different process for retrieving the stream (streams are identified with a "video_id" rather than "channel_id"), so the api doesn't work; however, I have been able to replicate what pafy does and retrieve the video info, such as stream urls. It's definitely still a work-in-progress, but I'll have to see if I can make something come of it.

**Comment by bigdav1178** [comment]
I've gotten a bit further using the YouTube variant as a template for YTTV. I had to rewrite the ./origin/origin_channels.py file, subbing chunks of my script in to return channel_list / stream_url; and using a test script, it appears to be working: I get the channel list (formatted in the same fashion as the original origin_channels.py script), and I get the stream url based on the origin_id.

My next hurdle is getting the channels to show up in the fHDHR web gui. The scan/import completes (command line / log says it found / imported the right number of channels), and using sqlitebrowser I can see the channel entries in ./data/cache/fhdhr.db; but the fHDHR web gui reports 0 channels.

Thoughts? Where should I be looking to troubleshoot this one?

**Comment by deathbybandaid** [comment]
I'm investigating an issue with this. Restarting after an intial import seems to help

**Comment by bigdav1178** [comment]
Hey, that did it! Thanks!

Off to my next hurdle...

**Comment by deathbybandaid** [comment]
Left myself a note to figure out what's going on with the channels not loadimg. Might just have the channels page try to reload from database if empty.

**Comment by deathbybandaid** [comment]
Let me know when you get to a comfortable point, and we could add it to the fHDHR umbrella

**Comment by bigdav1178** [comment]
Will do; it's not there yet, though. It is firing off the stream at this point; but I'm getting 403 errors, so I must need to pass some additional params with the auth. Also need to figure out how to have the functions use the same webdriver instance; I set it up as a global assignment, but it's still spawning multiple instances. A bit more work till its functional.

**Comment by deathbybandaid** [comment]
`self.fhdhr.web.session.get(url)` uses a requests session to preserve cookies and whatnot.

**Comment by deathbybandaid** [comment]
newest push to all varients fixes the channel loading issue

**Comment by bigdav1178** [comment]
Could be a while before I have anything functional. Part of those URL arguments that are missing is a decoded signature cipher; I can find modules that decode a Youtube video cipher, but they don't work with a Youtube TV URL. I may need to explore a different method to capturing the stream URLs (perhaps pulling them from the webdriver log, which has the decoded signature?).

**Comment by deathbybandaid** [comment]
Two of the varients use selenium/Firefox to scrape pages, if that helps

**Comment by bigdav1178** [comment]
I've been using selenium/Chrome so far. Late yesterday, I came across selenium-wire, and this may give me the request URL I need (which contains the decoded signature); but I need to tweak the config of it a bit - I'm getting socket.timeout's before getting the URL I'm looking for. Still more work to do; but progress, I believe.

**Comment by bigdav1178** [comment]
Using mitmproxy within seleniumwire has gotten me around the "socket.timeout" errors, and I'm finally able to grab URL strings with the decoded signature. The next thing I need to do is automate the selection of video quality (it's defaulting to 480p), so I get the beststream URLs; but that shouldn't be difficult.

My next question is this: how do we handle a stream that is split into separate URLs for audio and video? - ffmpeg to combine the a/v streams, then fHDHR serves the resulting ffmpeg output?

**Comment by deathbybandaid** [comment]
I had the same issue with split audio and video for a never completed varient, and that was too big of a roadblock to overcome

**Comment by bigdav1178** [comment]
I'm not ready to give up on it yet; I'll let you know if I manage to get something working.

**Comment by deathbybandaid** [comment]
if you can figure out how to get ffmpeg or vlc to handle split m3u8 audio/video, I just made some changes to `get_stream` to allow returning a dict, and I could code in some logic for using more than 1 input source

**Comment by deathbybandaid** [comment]
@bigdav1178 not overly helpful, but from an earlier version of fHDHR, I was playing with this for another varient

https://github.com/deathbybandaid/fHDHR_IHOP-KC-PrayerRoom/blob/dev/fHDHR/fHDHRweb/fHDHRdevice/watch.py#L81-L84

**Comment by deathbybandaid** [comment]
essentially this was supposed to use one url of video, and the other for audio,,, I just never got it to work

**Comment by bigdav1178** [comment]
No problem. Thanks for the help and suggestions; I will have to take a look into the changes you've made, as well as that other variant. I have seen ffmpeg examples to combine separate audio and video streams, so I will probably go down that path; however, when I tested the video-only stream, it failed to load the video - so I obviously have more work to go before I even get to that point.

In the interest of getting something working sooner rather than later, I've decided to fall back to my original method of streaming the content: livestreaming the desktop (chrome window) possibly via xvfb. It's a kludge; it's heavy on resource usage, and stream quality is dependent on the hardware used. But it is functional. I've been able to stream to vlc successfully using my own script. I'm in the process of cleaning up my existing origin_channels.py file (and likely origin_service.py), then integrating the commands from my test streaming script. My biggest hurdle right now is trying to get the webdriver to reuse the existing session, so it doesn't have to reauthenticate every time there's a channel change; I'm thinking that initiating my webdriver session in the origin_service file may resolve this. I'm also looking into suggestions using webdriver remote and session_id / command_executor. Hopefully one (or both) of these options pan out, and I can complete this version.

Once I get that version working, I plan to circle back to getting the separate a/v streams working. It'd be the much better solution (better quality stream and drastically less resource usage), but I also foresee it taking much longer to solve (if possible).
